From 3035236d8b7bfc615bd118ef018e028b9d8a38b4 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 10 Oct 2014 21:03:34 -0400
Subject: [block] blk-timeout: fix blk_add_timer

Message-id: <1412975015-5370-34-git-send-email-jmoyer@redhat.com>
Patchwork-id: 97464
O-Subject: [RHEL7 PATCH 33/34] blk-timeout: fix blk_add_timer
Bugzilla: 1146660
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Mike Snitzer <snitzer@redhat.com>

This is a backport of the following commit.  This resolves bug
1146660.

  commit 5e940aaa597c15e916618240ae5838864f36c91e
  Author: Ming Lei <ming.lei@canonical.com>
  Date:   Fri Sep 19 21:53:46 2014 +0800

    blk-timeout: fix blk_add_timer

    Commit 8cb34819cdd5d(blk-mq: unshared timeout handler) introduces
    blk-mq's own timeout handler, and removes following line:

     blk_queue_rq_timed_out(q, blk_mq_rq_timed_out);

    which then causes blk_add_timer() to bypass adding the timer,
    since blk-mq no longer has q->rq_timed_out_fn defined.

    This patch fixes the problem by bypassing the check for blk-mq,
    so that both request deadlines are still set and the rolling
    timer updated.

    Signed-off-by: Ming Lei <ming.lei@canonical.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-timeout.c b/block/blk-timeout.c
index 2fb1487..601d6fa 100644
--- a/block/blk-timeout.c
+++ b/block/blk-timeout.c
@@ -185,7 +185,8 @@ void blk_add_timer(struct request *req)
  struct request_queue *q = req->q;
  unsigned long expiry;
 
- if (!q->rq_timed_out_fn)
+ /* blk-mq has its own handler, so we don't need ->rq_timed_out_fn */
+ if (!q->mq_ops && !q->rq_timed_out_fn)
   return;
 
  BUG_ON(!list_empty(&req->timeout_list));
-- 
1.7.1